ynr/apps/people/models.py

```diff
@@ -57,6 +64,92 @@ class PersonImage(models.Model):
     objects = PersonImageManager()
 
 
+class PersonIdentifier(TimeStampedModel):
+    """
+    This is a model for storing "identifiers" for a person.
+
+    The most simple case is a URL to another website.
+
+    In this case the "value" is "@democlub" (or it could be
+    "https"//twitter.com/democlub").
+
+    Twitter uses "vanity URLs" as well as internal identifiers so we can
+    store the internal ID that looks like "362837635" on the internal_identifier
+    field.
+
+    An identifier value or internal identifier doesn't have to be resolvable
+    over HTTP. For example a phone number or snapchat handle are both valid
+    values.
+
+    """
+
+    person = models.ForeignKey(
+        "people.Person", related_name="tmp_person_identifiers"
+    )
+    value = models.CharField(
+        max_length=800,
+        help_text="An identifier e.g a URL or username provided by a 3rd party",
+    )
+    internal_identifier = models.CharField(
+        max_length=800,
+        help_text="An optional internal identifier from the 3rd party",
+        null=True,
+    )
+    value_type = models.CharField(
+        max_length=100,
+        help_text="A label for the type of value e.g. 'Twitter', 'Person blog'",
+    )
+    extra_data = JSONField(
+        help_text="""For storing any additional data against this field. 
+                     Used by bots, not humans.""",
+        null=True,
+    )
+
+    objects = PersonIdentifierQuerySet.as_manager()
+
+    class Meta:
+        unique_together = (
+            ("person", "value"),
+            ("person", "internal_identifier", "value_type"),
+            # TODO: Remove this.
+            # At the moemnt the version history can't deal with more than one
+            # value per value_type. This prevents creating duplicates, and
+            # therefore means we'll be able to merge and revert people.
+            # This constraint should be removed when the version history can
+            # support the data model we want
+            ("person", "value_type"),
+        )
+        ordering = ("value_type", "-modified")
+
+    def __str__(self):
+        return "{}: {} ({})".format(self.person_id, self.value_type, self.value)
+
+    @property
+    def get_value_type_html(self):
+        return self.value_type.title()
+
+    @property
+    def get_value_html(self):
+        url = None
+        text = "{}"
+
+        if self.value_type == "twitter":
+            if self.value.startswith("@"):
+                self.value = self.value[1:]
+            url = format_html("https://twitter.com/{}", self.value)
```

Comment:
Do we need to check if this is already a link?

i.e: if I've added "https://twitter.com/democlub/" and tagged it as "twitter" are we going to end up with "https://twitter.com/https://twitter.com/democlub/" ?

Reply:
I think this is really about form validation – I've not re-implemented this, but the previous set up had [form validation](https://github.com/DemocracyClub/yournextrepresentative/blob/84f3e5f355cb4efa7cbdf9291647b021db68a023/ynr/apps/people/forms.py#L161) for cleaning the entered username.

I guess we might have bots or script that import values, so it makes sense to clean it here too.